app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java

```diff
@@ -0,0 +1,306 @@
+                        e.printStackTrace();
+                        Log.e(TAG, "PhotoTester constructor -> Error decoding JSON with name " + fileName);
+                    }
+                    if(jsonPhotoDescription != null) {
+                        TestElement originalTest = new TestElement(photoBitmap, jsonPhotoDescription, fileName);
+
+                        //associate the relative bitmap to each alteration of the original test if there is any
+                        String[] alterationsFilenames = originalTest.getAlterationsNames();
+                        if(alterationsFilenames != null) {
+                            for(String alterationFilename : alterationsFilenames) {
+
+                                String alterationBitmapPath = dirPath + "/" + alterationFilename;
+                                Bitmap alterationBitmap = Utils.loadBitmapFromFile(alterationBitmapPath);
+                                //Utils.loadBitmapFromFile may return null
+                                if(alterationBitmap != null)
+                                    originalTest.setAlterationBitmap(alterationFilename, alterationBitmap);
+                                else
+                                    Log.i(TAG, "alteration image with name " + alterationFilename + " not found")
+                            }
+                        }
+                        testElements.add(originalTest);
+                    }
+                }
+            }
+        }
+    }
+
+
+    /**
+     * Elaborate tests using threads, stores a report in JSON format to testReport.txt inside the directory given on construction
+     * @return String in JSON format with the test's report, each object is a single test named with the filename and contains:
+     * ingredients, tags, notes, original photo name, recognized text, confidence and alterations (if any), each alteration
+     * contains tags, notes, recognized text, confidence
+     * @author Luca Moroldo (g3)
+     */
+    public String testAndReport() throws InterruptedException {
+        /*
+        This method will launch a task to evaluate each test, and uses:
+            - FixedThreadPool: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Executors.html#newFixedThreadPool-int-
+            - CounDownLatch: https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html
+
+        When all the tasks are completed, it saves a report in JSON format and return it as String.
+         */
+        final JSONObject fullJsonReport = new JSONObject();
+
+        int totalTestElements = testElements.size();
+        //countDownLatch allows to sync this thread with the end of all the single tests
+        CountDownLatch countDownLatch = new CountDownLatch(totalTestElements);
+
+        int max_concurrent_tasks = Runtime.getRuntime().availableProcessors();
+        //Define a thread executor that will run a maximum of 'max_concurrent_tasks' simultaneously
```

Comment:
Because this method will potentially run more that 200 tests, and doing this tests simultaneously using the available cores can reduce the time required